src/transpose.rs

```diff
@@ -0,0 +1,105 @@
+use std::thread;
+use std::time::Duration;
+
+use fluidsynth::synth::Synth;
+use log::info;
+
+use crate::notemap::NoteMap;
+use crate::synth::beep;
+
+#[derive(Copy, Clone, PartialEq)]
+enum TransposeCmd {
+    HalfStepUp,
+    HalfStepDown,
+    Direct(i32),
+    None,
+}
+
+// To set the transpose directly from a note we measure how far away from
+// the reference the note is. This is then used as the offset. We use high C
+// as the reference so we can fit soprano (High Bb -> -2), alto (Mid Eb -> -9),
+// tenor (Mid Bb, -> -14) and bari (Low Eb -> -21) within the reachable range.
+const TRANSPOSE_REFERENCE: i32 = 84;
+
+// Get the command associated with the given key state and volume.
+// For 'direct' transposition, the note must be played.
+fn get_cmd(key: u32, vol: i32, notemap: &NoteMap) -> TransposeCmd {
+    match (key, notemap.get_untransposed(&key), vol) {
+        (_, Some(note), v) if v > 10 => TransposeCmd::Direct(note),
+        (0x10000, None, _) => TransposeCmd::HalfStepUp,
```

Comment:
Can you define `R1` and `R3` as consts? i.e. `0x1_0000` and `0x40_0000`?

Reply:
Done